### PR TITLE
Replace slow `build_entity` and `build_from_entity` methods on cache data class with abstractmethods

### PR DESCRIPTION
**Pull request by FasterSpeeding**
### Summary
Replace slow `build_entity` and `build_from_entity` methods on cache data class with abstractmethods

### Checklist

- [ x] I have run `nox` and all the pipelines have passed.
- [ x] I have made unittests according to the code I have added/modified/deleted.

### CODE REVIEW

**Comment by FasterSpeeding** [comment]
i've not bothered to time this and compare that with the previous method but meh

File: hikari/utilities/cache.py [hunk 1]
```diff
@@ -368,20 +342,23 @@ class PrivateTextChannelData(BaseData[channels.PrivateTextChannel]):
...
 
-    @classmethod
-    def get_fields(cls) -> typing.Collection[str]:
-        return "id", "name", "last_message_id"
-
-    def build_entity(
-        self, target: typing.Type[channels.PrivateTextChannel], **kwargs: typing.Any
-    ) -> channels.PrivateTextChannel:
-        return super().build_entity(target, type=channels.ChannelType.PRIVATE_TEXT, **kwargs)
+    def build_entity(self, **kwargs: typing.Any) -> channels.PrivateTextChannel:
```
Comment (nekokatt) [inline hikari/utilities/cache.py:345]: isn't passing a dict each time equally inefficient?

Comment (FasterSpeeding) [reply hikari/utilities/cache.py:345]: we were doing that before and that's the only way to not violate that guy's law about substitution here and annoy mypy

**Comment by nekokatt** [comment]
> i've not bothered to time this and compare that with the previous method but meh

@FasterSpeeding so you have no clue if this runs twice as slow? I am not really sold on wanting to merge this if you don't even know what impact it has on the application. Merging that would be irresponsible.

**Comment by FasterSpeeding** [comment]
> 
> 
> > i've not bothered to time this and compare that with the previous method but meh
> 
> @FasterSpeeding so you have no clue if this runs twice as slow? I am not really sold on wanting to merge this if you don't even know what impact it has on the application. Merging that would be irresponsible.

if you want to time it then go ahead but i've got a migraine so i don't want to touch my pc again tonight

File: hikari/utilities/cache.py (hunk 1 above)
Comment (FasterSpeeding) [reply hikari/utilities/cache.py:345]: Liskov's substitution principle*

**Comment by nekokatt** [comment]
> > > i've not bothered to time this and compare that with the previous method but meh
> > 
> > 
> > @FasterSpeeding so you have no clue if this runs twice as slow? I am not really sold on wanting to merge this if you don't even know what impact it has on the application. Merging that would be irresponsible.
> 
> if you want to time it then go ahead but i've got a migraine so i don't want to touch my pc again tonight

ill trust your judgement on it, but please ensure you do check this, otherwise it completely defeats the object of doing anything if we just blindly take steps backwards randomly because we don't bother to be thorough.

**Pull request merged**